core/desugarMutual.ml

```diff
@@ -0,0 +1,153 @@
+open Sugartypes
+open Utility
+open SourceCode
+open Errors
+
+(* Desugars Mutual. *)
+(*
+ * mutual {
+ *   typename Foo = ...;
+ *   typename Bar = ...;
+ *   fun foo() {
+ *       ...
+ *   }
+ *
+ *  fun bar() {
+ *       ...
+ *   }
+ * } -->
+ * Typenames [Foo; Bar]
+ * Funs [foo; bar]
+ *
+ * Any other binding occurring in a mutual block is forbidden.
+ *)
+
+(* Simple check to see whether a function is recursive: checks whether the
+ * binder occurs within the free variables of the function body. *)
+let is_recursive bnd fnlit =
+  StringSet.mem (Binder.to_name bnd) (Freevars.funlit fnlit)
+
+
+(* Fun bindings must be lifted into `Funs if they are recursive. This is
+ * performed after the first pass. *)
+let lift_funs =
+object ((self : 'self_type))
+    inherit SugarTraversals.map as super
+    method! binding = fun b ->
+      let position = WithPos.pos b in
+      match WithPos.node b with
+      |  Fun (bndr, lin, (tvs, fnlit), location, dt) ->
+          let fnlit = self#funlit fnlit in
+          if is_recursive bndr fnlit then
+            WithPos.make ~pos:position (Funs [(bndr, lin, ((tvs, None), fnlit), location, dt, position)])
+          else
+            WithPos.make ~pos:position (Fun (bndr, lin, (tvs, fnlit), location, dt))
+      | _ -> super#binding b
+end
+
+let check_dups funs tys =
```

Comment:
There's a `has_duplicates` function in `utility.ml`. Would it be possible to rely on it instead of duplicating/reinventing the logic of duplicate checking?

Reply:
I considered this, but `has_duplicates` does not retain the information we need in order to provide an informative error message.

Reply:
Personally, I would be in favour of recovering this information when an error message is to be generated. This might require adding some new general-purpose list functions in `utility.ml`, but I would rather see new functionality being added rather than existing functionality being duplicated.